// @vkontakte/vkui/dist/components/RangeSlider/RangeSlider.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import _slicedToArray from "@babel/runtime/helpers/slicedToArray";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _defineProperty from "@babel/runtime/helpers/defineProperty";

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

import React, { Component, createRef } from 'react';
import Touch from '../Touch/Touch';
import getClassName from '../../helpers/getClassName';
import classNames from '../../lib/classNames';
import { precisionRound } from '../Slider/Slider';
import withPlatform from '../../hoc/withPlatform';
import { canUseDOM } from '../../lib/dom';
import { setRef } from '../../lib/utils';
import withAdaptivity from '../../hoc/withAdaptivity';

var RangeSlider = /*#__PURE__*/function (_Component) {
  _inherits(RangeSlider, _Component);

  var _super = _createSuper(RangeSlider);

  function RangeSlider(props) {
    var _this;

    _classCallCheck(this, RangeSlider);

    _this = _super.call(this, props);

    _defineProperty(_assertThisInitialized(_this), "isControlledOutside", void 0);

    _defineProperty(_assertThisInitialized(_this), "container", void 0);

    _defineProperty(_assertThisInitialized(_this), "thumbStart", void 0);

    _defineProperty(_assertThisInitialized(_this), "thumbEnd", void 0);

    _defineProperty(_assertThisInitialized(_this), "onStart", function (e) {
      var boundingRect = _this.container.getBoundingClientRect();

      _this.setState({
        containerWidth: boundingRect.width
      }, function () {
        var absolutePosition = _this.validateAbsolute(e.startX - boundingRect.left);

        var percentPosition = _this.absoluteToPecent(absolutePosition);

        var _this$calcPercentRang = _this.calcPercentRange(percentPosition),
            percentStart = _this$calcPercentRang.percentStart,
            percentEnd = _this$calcPercentRang.percentEnd;

        _this.onChange([_this.percentToValue(percentStart), _this.percentToValue(percentEnd)], e);

        if (_this.isControlledOutside) {
          _this.setState({
            startX: absolutePosition
          });
        } else {
          _this.setState({
            startX: absolutePosition,
            percentStart: percentStart,
            percentEnd: percentEnd
          });
        }
      });
    });

    _defineProperty(_assertThisInitialized(_this), "onMoveX", function (e) {
      var absolutePosition = _this.validateAbsolute(_this.state.startX + (e.shiftX || 0));

      var percentPosition = _this.absoluteToPecent(absolutePosition);

      var _this$calcPercentRang2 = _this.calcPercentRange(percentPosition),
          percentStart = _this$calcPercentRang2.percentStart,
          percentEnd = _this$calcPercentRang2.percentEnd;

      _this.onChange([_this.percentToValue(percentStart), _this.percentToValue(percentEnd)], e);

      if (!_this.isControlledOutside) {
        _this.setState({
          percentStart: percentStart,
          percentEnd: percentEnd
        });
      }

      e.originalEvent.preventDefault();
    });

    _defineProperty(_assertThisInitialized(_this), "getRef", function (container) {
      _this.container = container;
      setRef(container, _this.props.getRootRef);
    });

    _this.state = {
      startX: 0,
      percentStart: 0,
      percentEnd: 0,
      containerWidth: 0
    };
    _this.isControlledOutside = _this.props.hasOwnProperty('value');
    _this.thumbStart = /*#__PURE__*/createRef();
    _this.thumbEnd = /*#__PURE__*/createRef();
    return _this;
  }

  _createClass(RangeSlider, [{
    key: "onChange",
    value: function onChange(value, e) {
      this.props.onChange && this.props.onChange(value, e);
    }
  }, {
    key: "validateAbsolute",
    value: function validateAbsolute(absolute) {
      var res = Math.max(0, Math.min(absolute, this.state.containerWidth));

      if (this.props.step > 0) {
        var stepCount = (this.props.max - this.props.min) / this.props.step;
        var absStep = this.state.containerWidth / stepCount;
        res = Math.floor(res / absStep) * absStep;
      }

      return res;
    }
  }, {
    key: "validatePercent",
    value: function validatePercent(_ref) {
      var percentStart = _ref.percentStart,
          percentEnd = _ref.percentEnd;
      return {
        percentStart: Math.max(0, Math.min(percentStart, 100)),
        percentEnd: Math.max(0, Math.min(percentEnd, 100))
      };
    }
  }, {
    key: "absoluteToPecent",
    value: function absoluteToPecent(absolute) {
      return absolute * 100 / this.state.containerWidth;
    }
  }, {
    key: "calcPercentRange",
    value: function calcPercentRange(percent) {
      var _this$state = this.state,
          percentStart = _this$state.percentStart,
          percentEnd = _this$state.percentEnd;

      if (percentStart === 100) {
        return {
          percentStart: percent,
          percentEnd: percentEnd
        };
      } else if (percentEnd === 0) {
        return {
          percentEnd: percent,
          percentStart: percentStart
        };
      } else if (Math.abs(percentStart - percent) <= Math.abs(percentEnd - percent)) {
        return {
          percentStart: percent,
          percentEnd: percentEnd
        };
      } else {
        return {
          percentEnd: percent,
          percentStart: percentStart
        };
      }
    }
  }, {
    key: "percentToValue",
    value: function percentToValue(percent) {
      var res = percent * (this.props.max - this.props.min) / 100 + this.props.min;

      if (this.props.step > 0) {
        var stepFloatPart = "".concat(this.props.step).split('.')[1] || '';
        return precisionRound(res, stepFloatPart.length);
      }

      return res;
    }
  }, {
    key: "valueToPercent",
    value: function valueToPercent(_ref2) {
      var _ref3 = _slicedToArray(_ref2, 2),
          valueStart = _ref3[0],
          valueEnd = _ref3[1];

      return {
        percentStart: (valueStart - this.props.min) * 100 / (this.props.max - this.props.min),
        percentEnd: (valueEnd - this.props.min) * 100 / (this.props.max - this.props.min)
      };
    }
  }, {
    key: "componentDidMount",
    value: function componentDidMount() {
      var _this2 = this;

      if (canUseDOM) {
        var boundingRect = this.container.getBoundingClientRect();
        this.setState({
          containerWidth: boundingRect.width
        }, function () {
          _this2.setState(_this2.validatePercent(_this2.valueToPercent(_this2.value)));
        });
      }
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps) {
      if (this.isControlledOutside && prevProps.value !== this.props.value) {
        this.setState(this.validatePercent(this.valueToPercent(this.props.value)));
      }
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          className = _this$props.className,
          min = _this$props.min,
          max = _this$props.max,
          step = _this$props.step,
          value = _this$props.value,
          defaultValue = _this$props.defaultValue,
          onChange = _this$props.onChange,
          getRootRef = _this$props.getRootRef,
          platform = _this$props.platform,
          sizeY = _this$props.sizeY,
          restProps = _objectWithoutProperties(_this$props, ["className", "min", "max", "step", "value", "defaultValue", "onChange", "getRootRef", "platform", "sizeY"]);

      return /*#__PURE__*/React.createElement("div", _extends({}, restProps, {
        className: classNames(getClassName('Slider', platform), className, "Slider--sizeY-".concat(sizeY))
      }), /*#__PURE__*/React.createElement(Touch, {
        getRootRef: this.getRef,
        onStart: this.onStart,
        onMoveX: this.onMoveX,
        className: "Slider__in"
      }, /*#__PURE__*/React.createElement("div", {
        className: "Slider__dragger",
        style: {
          width: "".concat(this.state.percentEnd - this.state.percentStart, "%"),
          left: "".concat(this.state.percentStart, "%")
        }
      }, /*#__PURE__*/React.createElement("span", {
        className: classNames('Slider__thumb', 'Slider__thumb--start'),
        ref: this.thumbStart
      }), /*#__PURE__*/React.createElement("span", {
        className: classNames('Slider__thumb', 'Slider__thumb--end'),
        ref: this.thumbEnd
      }))));
    }
  }, {
    key: "value",
    get: function get() {
      if (this.isControlledOutside) {
        return this.props.value;
      } else if (this.props.hasOwnProperty('defaultValue')) {
        return this.props.defaultValue;
      } else {
        return [this.props.min, this.props.max];
      }
    }
  }]);

  return RangeSlider;
}(Component);

_defineProperty(RangeSlider, "defaultProps", {
  min: 0,
  max: 100,
  step: 0
});

export default withAdaptivity(withPlatform(RangeSlider), {
  sizeY: true
});
