// @vkontakte/vkui/dist/components/SelectMimicry/SelectMimicry.js
import _extends from "@babel/runtime/helpers/extends";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import React from 'react';
import classNames from '../../lib/classNames';
import { Icon24Dropdown, Icon20Dropdown } from '@vkontakte/icons';
import FormField from '../FormField/FormField';
import withAdaptivity, { SizeType } from '../../hoc/withAdaptivity';
import usePlatform from '../../hooks/usePlatform';
import { getClassName } from '../..';

var SelectMimicry = function SelectMimicry(_ref) {
  var _classNames;

  var className = _ref.className,
      tabIndex = _ref.tabIndex,
      placeholder = _ref.placeholder,
      children = _ref.children,
      align = _ref.align,
      getRootRef = _ref.getRootRef,
      multiline = _ref.multiline,
      disabled = _ref.disabled,
      onClick = _ref.onClick,
      sizeX = _ref.sizeX,
      sizeY = _ref.sizeY,
      restProps = _objectWithoutProperties(_ref, ["className", "tabIndex", "placeholder", "children", "align", "getRootRef", "multiline", "disabled", "onClick", "sizeX", "sizeY"]);

  var platform = usePlatform();
  return /*#__PURE__*/React.createElement(FormField, _extends({}, restProps, {
    tabIndex: disabled ? null : tabIndex,
    className: classNames(getClassName('Select', platform), 'Select--mimicry', (_classNames = {
      'Select--not-selected': !children,
      'Select--multiline': multiline,
      'Select--disabled': disabled
    }, _defineProperty(_classNames, "Select--align-".concat(align), !!align), _defineProperty(_classNames, "Select--sizeX--".concat(sizeX), !!sizeX), _defineProperty(_classNames, "Select--sizeY--".concat(sizeY), !!sizeY), _classNames), className),
    getRootRef: getRootRef,
    onClick: disabled ? null : onClick
  }), /*#__PURE__*/React.createElement("div", {
    className: "Select__container"
  }, /*#__PURE__*/React.createElement("div", {
    className: "Select__title"
  }, children || placeholder), sizeY === SizeType.COMPACT ? /*#__PURE__*/React.createElement(Icon20Dropdown, null) : /*#__PURE__*/React.createElement(Icon24Dropdown, null)));
};

SelectMimicry.defaultProps = {
  tabIndex: 0
};
export default withAdaptivity(SelectMimicry, {
  sizeX: true,
  sizeY: true
});
