// @vkontakte/vkui/dist/components/Checkbox/Checkbox.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import React from 'react';
import Tappable, { ACTIVE_EFFECT_DELAY } from '../Tappable/Tappable';
import getClassName from '../../helpers/getClassName';
import classNames from '../../lib/classNames';
import { IOS, VKCOM } from '../../lib/platform';
import { Icon20CheckBoxOn, Icon20CheckBoxOff, Icon24CheckBoxOn, Icon24CheckBoxOff } from '@vkontakte/icons';
import usePlatform from '../../hooks/usePlatform';
import withAdaptivity, { SizeType } from '../../hoc/withAdaptivity';
import Text from '../Typography/Text/Text';
import Headline from '../Typography/Headline/Headline';
export var Checkbox = function Checkbox(_ref) {
  var children = _ref.children,
      className = _ref.className,
      style = _ref.style,
      getRootRef = _ref.getRootRef,
      getRef = _ref.getRef,
      sizeY = _ref.sizeY,
      restProps = _objectWithoutProperties(_ref, ["children", "className", "style", "getRootRef", "getRef", "sizeY"]);

  var platform = usePlatform();
  var ContentComponent = platform === VKCOM || sizeY === SizeType.COMPACT ? Text : Headline;
  return /*#__PURE__*/React.createElement(Tappable, {
    Component: "label",
    className: classNames(getClassName('Checkbox', platform), className, "Checkbox--sizeY-".concat(sizeY)),
    style: style,
    disabled: restProps.disabled,
    activeEffectDelay: platform === IOS ? 100 : ACTIVE_EFFECT_DELAY,
    getRootRef: getRootRef
  }, /*#__PURE__*/React.createElement("input", _extends({}, restProps, {
    type: "checkbox",
    className: "Checkbox__input",
    ref: getRef
  })), /*#__PURE__*/React.createElement("div", {
    className: "Checkbox__container"
  }, /*#__PURE__*/React.createElement("div", {
    className: "Checkbox__icon Checkbox__icon--on"
  }, sizeY === SizeType.COMPACT || platform === VKCOM ? /*#__PURE__*/React.createElement(Icon20CheckBoxOn, null) : /*#__PURE__*/React.createElement(Icon24CheckBoxOn, null)), /*#__PURE__*/React.createElement("div", {
    className: "Checkbox__icon Checkbox__icon--off"
  }, sizeY === SizeType.COMPACT || platform === VKCOM ? /*#__PURE__*/React.createElement(Icon20CheckBoxOff, null) : /*#__PURE__*/React.createElement(Icon24CheckBoxOff, null)), /*#__PURE__*/React.createElement(ContentComponent, {
    weight: "regular",
    className: "Checkbox__content"
  }, children)));
};
export default withAdaptivity(Checkbox, {
  sizeY: true
});
