// @vkontakte/vkui/dist/components/Input/Input.js
import _extends from "@babel/runtime/helpers/extends";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import React from 'react';
import getClassName from '../../helpers/getClassName';
import classNames from '../../lib/classNames';
import FormField from '../FormField/FormField';
import withAdaptivity from '../../hoc/withAdaptivity';
import usePlatform from '../../hooks/usePlatform';

var Input = function Input(_ref) {
  var align = _ref.align,
      getRef = _ref.getRef,
      className = _ref.className,
      getRootRef = _ref.getRootRef,
      sizeY = _ref.sizeY,
      restProps = _objectWithoutProperties(_ref, ["align", "getRef", "className", "getRootRef", "sizeY"]);

  var platform = usePlatform();
  return /*#__PURE__*/React.createElement(FormField, {
    className: classNames(getClassName('Input', platform), className, _defineProperty({}, "Input--".concat(align), !!align), "Input--sizeY-".concat(sizeY)),
    getRootRef: getRootRef
  }, /*#__PURE__*/React.createElement("input", _extends({}, restProps, {
    className: "Input__el",
    ref: getRef
  })));
};

Input.defaultProps = {
  type: 'text'
};
export default withAdaptivity(Input, {
  sizeY: true
});
