// @vkontakte/vkui/dist/components/Banner/Banner.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import React from 'react';
import getClassName from '../../helpers/getClassName';
import classNames from '../../lib/classNames';
import usePlatform from '../../hooks/usePlatform';
import { ANDROID, IOS, VKCOM } from '../../lib/platform';
import { Icon24Chevron, Icon24DismissSubstract, Icon24DismissDark, Icon24Cancel } from '@vkontakte/icons';
import Tappable from '../Tappable/Tappable';
import Headline from '../Typography/Headline/Headline';
import Caption from '../Typography/Caption/Caption';
import Text from '../Typography/Text/Text';
import { hasReactNode } from '../../lib/utils';
import Title from '../Typography/Title/Title';

function renderHeader(_ref) {
  var size = _ref.size,
      header = _ref.header;

  switch (size) {
    case 's':
      return /*#__PURE__*/React.createElement(Headline, {
        weight: "medium",
        className: "Banner__header"
      }, header);

    case 'm':
      return /*#__PURE__*/React.createElement(Title, {
        level: "2",
        weight: "medium",
        className: "Banner__header"
      }, header);
  }
}

function renderSubheader(_ref2) {
  var size = _ref2.size,
      subheader = _ref2.subheader;

  switch (size) {
    case 's':
      return /*#__PURE__*/React.createElement(Caption, {
        level: "1",
        weight: "regular",
        className: "Banner__subheader"
      }, subheader);

    case 'm':
      return /*#__PURE__*/React.createElement(Text, {
        weight: "regular",
        className: "Banner__subheader"
      }, subheader);
  }
}

var Banner = function Banner(props) {
  var platform = usePlatform();

  var className = props.className,
      mode = props.mode,
      imageTheme = props.imageTheme,
      size = props.size,
      before = props.before,
      asideMode = props.asideMode,
      header = props.header,
      subheader = props.subheader,
      text = props.text,
      children = props.children,
      background = props.background,
      actions = props.actions,
      onDismiss = props.onDismiss,
      restProps = _objectWithoutProperties(props, ["className", "mode", "imageTheme", "size", "before", "asideMode", "header", "subheader", "text", "children", "background", "actions", "onDismiss"]);

  var InnerComponent = asideMode === 'expand' ? Tappable : 'div';
  return /*#__PURE__*/React.createElement("div", _extends({}, restProps, {
    className: classNames(getClassName('Banner', platform), "Banner--md-".concat(mode), "Banner--sz-".concat(size), {
      'Banner--inverted': mode === 'image' && imageTheme === 'dark'
    }, className)
  }), /*#__PURE__*/React.createElement(InnerComponent, {
    className: "Banner__in"
  }, mode === 'image' && background && /*#__PURE__*/React.createElement("div", {
    className: "Banner__bg"
  }, background), before && /*#__PURE__*/React.createElement("div", {
    className: "Banner__before"
  }, before), /*#__PURE__*/React.createElement("div", {
    className: "Banner__content"
  }, hasReactNode(header) && renderHeader({
    size: size,
    header: header
  }), hasReactNode(subheader) && renderSubheader({
    size: size,
    subheader: subheader
  }), hasReactNode(text) && /*#__PURE__*/React.createElement(Text, {
    weight: "regular",
    className: "Banner__text"
  }, text), actions && /*#__PURE__*/React.createElement("div", {
    className: "Banner__actions"
  }, actions)), asideMode === 'expand' && /*#__PURE__*/React.createElement("div", {
    className: "Banner__expand"
  }, /*#__PURE__*/React.createElement(Icon24Chevron, null)), asideMode === 'dismiss' && /*#__PURE__*/React.createElement("div", {
    className: "Banner__dismiss"
  }, /*#__PURE__*/React.createElement("div", {
    className: "Banner__dismissIcon",
    onClick: onDismiss
  }, (platform === ANDROID || platform === VKCOM) && /*#__PURE__*/React.createElement(Icon24Cancel, null), platform === IOS && (mode === 'image' ? /*#__PURE__*/React.createElement(Icon24DismissDark, null) : /*#__PURE__*/React.createElement(Icon24DismissSubstract, null))))));
};

Banner.defaultProps = {
  mode: 'tint',
  size: 's',
  imageTheme: 'dark'
};
export default Banner;
