// @vkontakte/vkui/dist/components/Switch/Switch.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import React from 'react';
import getClassName from '../../helpers/getClassName';
import classNames from '../../lib/classNames';
import usePlatform from '../../hooks/usePlatform';
import withAdaptivity from '../../hoc/withAdaptivity';

var Switch = function Switch(_ref) {
  var style = _ref.style,
      className = _ref.className,
      getRef = _ref.getRef,
      getRootRef = _ref.getRootRef,
      sizeY = _ref.sizeY,
      restProps = _objectWithoutProperties(_ref, ["style", "className", "getRef", "getRootRef", "sizeY"]);

  var platform = usePlatform();
  return /*#__PURE__*/React.createElement("label", {
    className: classNames(getClassName('Switch', platform), className, "Switch--sizeY-".concat(sizeY)),
    style: style,
    ref: getRootRef
  }, /*#__PURE__*/React.createElement("input", _extends({}, restProps, {
    type: "checkbox",
    className: "Switch__self",
    ref: getRef
  })), /*#__PURE__*/React.createElement("span", {
    className: "Switch__pseudo"
  }));
};

export default withAdaptivity(Switch, {
  sizeY: true
});
