// @vkontakte/vkui/dist/components/RichCell/RichCell.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import React from 'react';
import classNames from '../../lib/classNames';
import usePlatform from '../../hooks/usePlatform';
import getClassName from '../../helpers/getClassName';
import Tappable from '../Tappable/Tappable';
import { hasReactNode } from '../../lib/utils';
import Text from '../Typography/Text/Text';
import Caption from '../Typography/Caption/Caption';
import withAdaptivity from '../../hoc/withAdaptivity';

var RichCell = function RichCell(_ref) {
  var children = _ref.children,
      text = _ref.text,
      caption = _ref.caption,
      before = _ref.before,
      after = _ref.after,
      bottom = _ref.bottom,
      actions = _ref.actions,
      multiline = _ref.multiline,
      className = _ref.className,
      Component = _ref.Component,
      onClick = _ref.onClick,
      sizeY = _ref.sizeY,
      restProps = _objectWithoutProperties(_ref, ["children", "text", "caption", "before", "after", "bottom", "actions", "multiline", "className", "Component", "onClick", "sizeY"]);

  var platform = usePlatform();
  var RootComponent = restProps.disabled ? Component : Tappable;
  Component = restProps.disabled ? undefined : Component;
  var props = restProps;

  if (!restProps.disabled) {
    props.Component = restProps.href ? 'a' : Component;
    props.onClick = onClick;
  }

  return /*#__PURE__*/React.createElement(RootComponent, _extends({}, props, {
    className: classNames(className, getClassName('RichCell', platform), {
      'RichCell--mult': multiline
    }, "RichCell--sizeY-".concat(sizeY))
  }), before, /*#__PURE__*/React.createElement("div", {
    className: "RichCell__in"
  }, /*#__PURE__*/React.createElement("div", {
    className: "RichCell__top"
  }, after, /*#__PURE__*/React.createElement(Text, {
    weight: "medium",
    className: "RichCell__content"
  }, /*#__PURE__*/React.createElement("div", {
    className: "RichCell__children"
  }, children), hasReactNode(after) && /*#__PURE__*/React.createElement("div", {
    className: "RichCell__after"
  }, after)), hasReactNode(text) && /*#__PURE__*/React.createElement(Text, {
    weight: "regular",
    className: "RichCell__text"
  }, text), hasReactNode(caption) && /*#__PURE__*/React.createElement(Caption, {
    level: "1",
    weight: "regular",
    className: "RichCell__caption"
  }, caption), (hasReactNode(bottom) || hasReactNode(actions)) && /*#__PURE__*/React.createElement("div", {
    className: "RichCell__bottom"
  }, bottom, hasReactNode(actions) && /*#__PURE__*/React.createElement("div", {
    className: "RichCell__actions"
  }, actions)))));
};

RichCell.defaultProps = {
  Component: 'div'
};
export default withAdaptivity(RichCell, {
  sizeY: true
});
