// @vkontakte/vkui/dist/components/Radio/Radio.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import React from 'react';
import Tappable, { ACTIVE_EFFECT_DELAY } from '../Tappable/Tappable';
import getClassName from '../../helpers/getClassName';
import classNames from '../../lib/classNames';
import { IOS, VKCOM } from '../../lib/platform';
import usePlatform from '../../hooks/usePlatform';
import withAdaptivity, { SizeType } from '../../hoc/withAdaptivity';
import { hasReactNode } from '../../lib/utils';
import Caption from '../../components/Typography/Caption/Caption';
import Headline from '../Typography/Headline/Headline';
import Text from '../Typography/Text/Text';

var Radio = function Radio(props) {
  var children = props.children,
      description = props.description,
      style = props.style,
      className = props.className,
      getRef = props.getRef,
      getRootRef = props.getRootRef,
      sizeY = props.sizeY,
      restProps = _objectWithoutProperties(props, ["children", "description", "style", "className", "getRef", "getRootRef", "sizeY"]);

  var platform = usePlatform();
  var ContentComponent = platform === VKCOM || sizeY === SizeType.COMPACT ? Text : Headline;
  var descriptionLevel = platform === VKCOM || sizeY === SizeType.COMPACT ? '2' : '1';
  return /*#__PURE__*/React.createElement(Tappable, {
    Component: "label",
    style: style,
    className: classNames(getClassName('Radio', platform), className, "Radio--sizeY-".concat(sizeY)),
    activeEffectDelay: platform === IOS ? 100 : ACTIVE_EFFECT_DELAY,
    disabled: restProps.disabled,
    getRootRef: getRootRef
  }, /*#__PURE__*/React.createElement("input", _extends({}, restProps, {
    type: "radio",
    className: "Radio__input",
    ref: getRef
  })), /*#__PURE__*/React.createElement("div", {
    className: "Radio__container"
  }, /*#__PURE__*/React.createElement("div", {
    className: "Radio__icon"
  }), /*#__PURE__*/React.createElement(ContentComponent, {
    weight: "regular",
    className: "Radio__content"
  }, /*#__PURE__*/React.createElement("div", {
    className: "Radio__children"
  }, children), hasReactNode(description) && /*#__PURE__*/React.createElement(Caption, {
    level: descriptionLevel,
    weight: "regular",
    className: "Radio__description"
  }, description))));
};

export default withAdaptivity(Radio, {
  sizeY: true
});
