// @vkontakte/vkui/dist/components/ActionSheetItem/ActionSheetItem.js
import _extends from "@babel/runtime/helpers/extends";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import React, { useContext } from 'react';
import classNames from '../../lib/classNames';
import getClassName from '../../helpers/getClassName';
import Tappable from '../Tappable/Tappable';
import usePlatform from '../../hooks/usePlatform';
import { hasReactNode } from '../../lib/utils';
import Subhead from '../Typography/Subhead/Subhead';
import Title from '../Typography/Title/Title';
import Text from '../Typography/Text/Text';
import { ANDROID, VKCOM } from '../../lib/platform';
import { Icon16Done, Icon24Done } from '@vkontakte/icons';
import { ActionSheetContext } from '../ActionSheet/ActionSheetContext';
import Caption from '../Typography/Caption/Caption';
import withAdaptivity, { SizeType } from '../../hoc/withAdaptivity';

var ActionSheetItem = function ActionSheetItem(_ref) {
  var _classNames;

  var className = _ref.className,
      children = _ref.children,
      autoclose = _ref.autoclose,
      mode = _ref.mode,
      meta = _ref.meta,
      subtitle = _ref.subtitle,
      before = _ref.before,
      isLast = _ref.isLast,
      selectable = _ref.selectable,
      value = _ref.value,
      name = _ref.name,
      checked = _ref.checked,
      defaultChecked = _ref.defaultChecked,
      onChange = _ref.onChange,
      onClick = _ref.onClick,
      sizeY = _ref.sizeY,
      restProps = _objectWithoutProperties(_ref, ["className", "children", "autoclose", "mode", "meta", "subtitle", "before", "isLast", "selectable", "value", "name", "checked", "defaultChecked", "onChange", "onClick", "sizeY"]);

  var platform = usePlatform();

  var _useContext = useContext(ActionSheetContext),
      onItemClick = _useContext.onItemClick,
      isDesktop = _useContext.isDesktop;

  var Component = 'div';

  if (restProps.href) {
    Component = 'a';
  } else if (selectable) {
    Component = 'label';
  }

  var isCompact = hasReactNode(subtitle) || hasReactNode(meta) || selectable;
  return /*#__PURE__*/React.createElement(Tappable, _extends({}, restProps, {
    onClick: onItemClick(onClick, autoclose),
    className: classNames(getClassName('ActionSheetItem', platform), "ActionSheetItem--".concat(mode), (_classNames = {
      'ActionSheetItem--compact': isCompact,
      'ActionSheetItem--desktop': isDesktop
    }, _defineProperty(_classNames, "ActionSheetItem--sizeY-".concat(sizeY), sizeY === SizeType.COMPACT), _defineProperty(_classNames, 'ActionSheetItem--withSubtitle', hasReactNode(subtitle)), _classNames), className),
    Component: Component
  }), hasReactNode(before) && /*#__PURE__*/React.createElement("div", {
    className: "ActionSheetItem__before"
  }, before), /*#__PURE__*/React.createElement("div", {
    className: "ActionSheetItem__container"
  }, /*#__PURE__*/React.createElement("div", {
    className: "ActionSheetItem__content"
  }, sizeY === SizeType.COMPACT ? /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Text, {
    weight: mode === 'cancel' ? 'medium' : 'regular',
    className: "ActionSheetItem__children"
  }, children), hasReactNode(meta) && /*#__PURE__*/React.createElement(Text, {
    weight: "regular",
    className: "ActionSheetItem__meta"
  }, meta)) : /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(Title, {
    weight: mode === 'cancel' ? 'medium' : 'regular',
    level: isCompact || hasReactNode(before) || platform === ANDROID ? '3' : '2',
    className: "ActionSheetItem__children"
  }, children), hasReactNode(meta) && /*#__PURE__*/React.createElement(Title, {
    weight: "regular",
    level: isCompact || hasReactNode(before) || platform === ANDROID ? '3' : '2',
    className: "ActionSheetItem__meta"
  }, meta))), hasReactNode(subtitle) && (sizeY === SizeType.COMPACT ? /*#__PURE__*/React.createElement(Caption, {
    weight: "regular",
    className: "ActionSheetItem__subtitle",
    level: "1"
  }, subtitle) : /*#__PURE__*/React.createElement(Subhead, {
    weight: "regular",
    className: "ActionSheetItem__subtitle"
  }, subtitle))), selectable && /*#__PURE__*/React.createElement("div", {
    className: "ActionSheetItem__after"
  }, /*#__PURE__*/React.createElement("input", {
    type: "radio",
    className: "ActionSheetItem__radio",
    name: name,
    value: value,
    onChange: onChange,
    defaultChecked: defaultChecked,
    checked: checked,
    disabled: restProps.disabled
  }), /*#__PURE__*/React.createElement("div", {
    className: "ActionSheetItem__marker"
  }, platform === VKCOM ? /*#__PURE__*/React.createElement(Icon24Done, null) : /*#__PURE__*/React.createElement(Icon16Done, null))));
};

ActionSheetItem.defaultProps = {
  mode: 'default'
};
export default withAdaptivity(ActionSheetItem, {
  sizeY: true
});
