// @vkontakte/vkui/dist/components/File/File.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import React, { useRef } from 'react';
import getClassName from '../../helpers/getClassName';
import Button from '../Button/Button';
import classNames from '../../lib/classNames';
import usePlatform from '../../hooks/usePlatform';
import { setRef } from '../../lib/utils';

var File = function File(props) {
  var children = props.children,
      align = props.align,
      controlSize = props.controlSize,
      mode = props.mode,
      stretched = props.stretched,
      before = props.before,
      className = props.className,
      style = props.style,
      getRef = props.getRef,
      getRootRef = props.getRootRef,
      _onClick = props.onClick,
      restProps = _objectWithoutProperties(props, ["children", "align", "controlSize", "mode", "stretched", "before", "className", "style", "getRef", "getRootRef", "onClick"]);

  var platform = usePlatform();
  var inputRef = useRef();

  var getInputRef = function getInputRef(element) {
    inputRef.current = element;
    setRef(element, getRef);
  };

  return /*#__PURE__*/React.createElement(Button, {
    align: align,
    className: classNames(getClassName('File', platform), className),
    stretched: stretched,
    mode: mode,
    size: controlSize,
    before: before,
    style: style,
    getRootRef: getRootRef,
    disabled: restProps.disabled,
    onClick: function onClick(e) {
      inputRef.current.click();
      _onClick && _onClick(e);
    }
  }, /*#__PURE__*/React.createElement("input", _extends({}, restProps, {
    className: "File__input",
    type: "file",
    ref: getInputRef
  })), children);
};

File.defaultProps = {
  children: 'Выберите файл',
  align: 'left'
};
export default File;
