// @vkontakte/vkui/dist/components/Tabs/Tabs.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import React from 'react';
import getClassName from '../../helpers/getClassName';
import classNames from '../../lib/classNames';
import usePlatform from '../../hooks/usePlatform';
import { ANDROID, VKCOM } from '../../lib/platform';
import withAdaptivity from '../../hoc/withAdaptivity';

var Tabs = function Tabs(_ref) {
  var className = _ref.className,
      children = _ref.children,
      style = _ref.style,
      mode = _ref.mode,
      getRootRef = _ref.getRootRef,
      sizeX = _ref.sizeX,
      restProps = _objectWithoutProperties(_ref, ["className", "children", "style", "mode", "getRootRef", "sizeX"]);

  var platform = usePlatform();

  if ((platform === ANDROID || platform === VKCOM) && mode === 'segmented') {
    mode = 'default';
  }

  return /*#__PURE__*/React.createElement("div", _extends({}, restProps, {
    ref: getRootRef,
    className: classNames(getClassName('Tabs', platform), "Tabs--".concat(mode), "Tabs--sizeX-".concat(sizeX), className),
    style: style
  }), /*#__PURE__*/React.createElement("div", {
    className: "Tabs__in"
  }, children));
};

Tabs.defaultProps = {
  mode: 'default'
};
export default withAdaptivity(Tabs, {
  sizeX: true
});
