// @vkontakte/vkui/dist/components/Typography/Title/Title.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import React from 'react';
import usePlatform from '../../../hooks/usePlatform';
import classNames from '../../../lib/classNames';
import getClassName from '../../../helpers/getClassName';
import { ANDROID } from '../../../lib/platform';
import Headline from '../Headline/Headline';

var getComponent = function getComponent(level) {
  if (!level) {
    return 'div';
  }

  return 'h' + level;
};

var getAndroidTitleWeight = function getAndroidTitleWeight(level, weight) {
  if (level === '3') {
    return weight === 'regular' ? weight : 'medium';
  }

  if (level === '2' && weight === 'semibold') {
    return 'medium';
  }

  if (weight === 'heavy') {
    return 'bold';
  }

  return weight;
};

var Title = function Title(_ref) {
  var children = _ref.children,
      className = _ref.className,
      weight = _ref.weight,
      level = _ref.level,
      Component = _ref.Component,
      restProps = _objectWithoutProperties(_ref, ["children", "className", "weight", "level", "Component"]);

  var platform = usePlatform();
  var TitleComponent = Component || getComponent(level);
  var titleWeight = platform === ANDROID ? getAndroidTitleWeight(level, weight) : weight;

  if (platform === ANDROID && level === '3') {
    return /*#__PURE__*/React.createElement(Headline, _extends({
      Component: TitleComponent
    }, restProps, {
      weight: titleWeight,
      className: className
    }), children);
  }

  return /*#__PURE__*/React.createElement(TitleComponent, _extends({}, restProps, {
    className: classNames(getClassName('Title', platform), "Title--w-".concat(titleWeight), "Title--l-".concat(level), className)
  }), children);
};

export default Title;
