// @vkontakte/vkui/dist/components/CellButton/CellButton.js
import _extends from "@babel/runtime/helpers/extends";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import React from 'react';
import getClassName from '../../helpers/getClassName';
import classNames from '../../lib/classNames';
import usePlatform from '../../hooks/usePlatform';
import SimpleCell from '../SimpleCell/SimpleCell';

var CellButton = function CellButton(_ref) {
  var className = _ref.className,
      centered = _ref.centered,
      mode = _ref.mode,
      restProps = _objectWithoutProperties(_ref, ["className", "centered", "mode"]);

  var platform = usePlatform();
  return /*#__PURE__*/React.createElement(SimpleCell, _extends({}, restProps, {
    className: classNames(getClassName('CellButton', platform), "CellButton--".concat(mode), _defineProperty({}, 'CellButton--centered', centered), className)
  }));
};

CellButton.defaultProps = {
  mode: 'primary',
  Component: 'button',
  centered: false,
  stopPropagation: true
};
export default CellButton;
